### PR TITLE
feat: multiple new features

### PR DESCRIPTION
**Pull request by pimlie**
- refactor(blueprint): define proper context and always pass that as first arg
- feat(blueprint): add basic support for multiple ids/configs per blueprint
- feat(blueprint): template definitions can be a function
- feat(docs): support dynamic loaded configs per locale
- feat(docs): support multiple folders for a blueprint
- feat(docs): support multiple ids per blueprint
- feat(docs): define sidebar items with glob/regex
- feat(docs): add custom (ie non nuxt/press page) items to sidebar
- feat(docs): support source folder aliasing
- test(docs): add more unit/e2e tests

As its probably best to show by example, have a look at the [nuxt.press.json](https://github.com/nuxt/press/blob/c3ed690d73ea4bb5fcc76cc7b2a87ddaef313de5/src/blueprints/docs/test/fixtures/full/nuxt.press.json) of the full e2e test

A lot of these changes were incorporated by trying to get nuxt/press to run with the nuxt-docs repo. We are probably not fully there yet, but I am confident we are at 90+% if not more.

The biggest change that affects all 3 modules is the refactor to a proper blueprint context. From now on, you should _always_ use this context and never refer to `this.$press` anymore. All functions within a blueprint (should) receive the context as first argument, except for `api` and `source` functions.
 The context contains a prop `rootOptions` which refers for nuxt/press to `this.$press` and a prop `options` which refers to the options of the current blueprint config (this is to support running a blueprint with multiple configs).

### Todo's
I have left some todo's in the code I'd appreciate both your help with. Especially the reactivity issues are quite annoying. I have now fixed that by using an event in [`layouts/docs.vue`](https://github.com/nuxt/press/blob/c3ed690d73ea4bb5fcc76cc7b2a87ddaef313de5/src/blueprints/docs/templates/layouts/docs.vue) and returning empty placeholders in [`mixins/docs.js`](https://github.com/nuxt/press/blob/c3ed690d73ea4bb5fcc76cc7b2a87ddaef313de5/src/blueprints/docs/templates/mixins/docs.js) but we should fix this properly. The issue is eg that the sidebar uses the docs mixin to retrieve it options, but the sidebar itself is defined in the layout. So when `this.$docs` changes, the observers trigger all at the same time and before the layout has had time to remove the sidebar, the sidebar itself already tried to update itself with wrong info (possibly resulting in an error without those empty placeholders)

### $press.get
~~We should remove the functionality of retrieving source files ourselves over http. We can just leverage webpack to do this for us just like I am doing now for the [doc configs](https://github.com/nuxt/press/blob/c3ed690d73ea4bb5fcc76cc7b2a87ddaef313de5/src/blueprints/docs/templates/plugins/press.docs.js#L50)~~
This is now impemented for anythingn stored in `press/static/sources` which has api url `/api/source/`
[Edit by @galvez]: the reason we're not removing `press.$get` is to retain the ability to do serverMiddleware overrides.

### routes
Ah the routes, I forgot the routes! They are as less-greedy as possible now, this means we could use different blueprint configurations for eg `/docs/(nl|en)` then for `/docs/(pt|br)`. This was because I noticed on nuxt-docs there are some languages which are not available on the website yet, eg we could add them now with a custom layout / styling.

@galvez Although I have added extensive testing I am sure I broke something. Sorry not sorry ;)
@Atinux @alexchopin Here is an example config if you wanna try this branch of nuxt/press on nuxt-docs

<details>
<summary>nuxt.press.json</summary>

```js
{
  "docs": {
    "dir": [
      "en",
      "pt-BR",
      "zh"
    ],
    "prefix": "/",
    "title": "Nuxt Docs",
    "search": true,
    "localizedSidebar": false,
    "configPerLocale": true,
    "nav": [
      {
        "Guide": "/en/guide/"
      },
      {
        "API": "/en/api/"
      }
    ],
    "sidebar": {
      "/api/": [
        {
          "title": "Essential",
          "children": [
            "/api/context"
          ]
        },
        {
          "title": "Pages",
          "children": [
            "/api/$",
            "/api/pages-*"
          ]
        },
        {
          "title": "Components",
          "children": [
            "/api/components-nu*",
            "/api/components-no*"
          ]
        },
        {
          "title": "Utils",
          "children": [
            "/api/$nuxt/"
          ]
        },
        {
          "title": "Configuration",
          "children": [
            "/api/configuration-*"
          ]
        },
        {
          "title": "Programmatically",
          "children": [
            "/api/nuxt*"
          ]
        },
        {
          "title": "Internals",
          "children": [
            "/api/internals*"
          ]
        }
      ],
      "/guide/": [
        {
          "title": "Prologue",
          "children": [
            "/guide/$",
            "/guide/contribution-guide",
            [
              "/guide/release-notes",
              "Release Notes"
            ],
            "/guide/upgrading"
          ]
        },
        {
          "title": "Getting Started",
          "children": [
            "/guide/installation",
            "/guide/directory-structure",
            "/guide/configuration",
            "/guide/routing"
          ]
        },
        {
          "title": "Resources",
          "children": [
            "/guide/themes"
          ]
        }
      ]
    }
  },
  "i18n": {
    "locales": [
      {
        "code": "en",
        "name": "English"
      },
      {
        "code": "pt-BR",
        "name": "Portuguese"
      },
      {
        "code": "zh",
        "name": "Chinese"
      }
    ]
  }
}
```
</details>

### CODE REVIEW

File: src/blueprints/docs/templates/components/topic.vue [hunk 1]
```diff
@@ -31,7 +31,11 @@ export default {
       { property: 'og:site_name', content: this.$docs.title }
     ]
 
-    if (this.meta.description) {
+    if (this.$press.docs.image) {
+      meta.push({ property: 'og:image', content: this.$press.docs.image })
```
Comment (pimlie) [inline src/blueprints/docs/templates/components/topic.vue]:
```suggestion
      meta.push({ property: 'og:image', content: this.$docs.image })
```
`this.$docs` is the ref to the docs blueprint config for the current pressId, see the docs mixin

File: src/blueprints/docs/templates/components/topic.vue [hunk 2]
```diff
@@ -31,7 +31,11 @@ export default {
       { property: 'og:site_name', content: this.$docs.title }
     ]
 
-    if (this.meta.description) {
+    if (this.$press.docs.image) {
```
Comment (pimlie) [inline src/blueprints/docs/templates/components/topic.vue]:
```suggestion
    if (this.$docs.image) {
```

File: src/blueprints/common/index.js [hunk 3]
```diff
@@ -125,10 +125,21 @@ export default {
       }
     },
     async done (context) {
-      if (!this.nuxt.options.dev) {
+      if (!context.rootOptions.dev) {
```
Comment (pimlie) [inline src/blueprints/common/index.js]: Whats actually supposed to be the difference between rootOptions.dev and nuxt.options.dev? Cause atm they are the same

Comment (galvez) [reply src/blueprints/common/index.js]: I thought we were supposed to always use the context object now?

Comment (pimlie) [reply src/blueprints/common/index.js]: That applies to the use of eg `this.$press`. Nuxt options should still always be the same for any blueprint root or blueprint config.

File: src/index.js [hunk 4]
```diff
@@ -70,6 +70,10 @@ export default async function NuxtPressModule (options) {
     next()
   })
 
+  this.extendBuild((config, { isClient }) => {
+
```
Comment (galvez) [inline src/index.js]: @pimlie this can be removed right?

Comment (pimlie) [reply src/index.js]: yeah, sry. Must be a leftover from testing something

**Pull request merged**